website/source/docs/platform/k8s/service-sync.html.md

```diff
@@ -0,0 +1,274 @@
+secret.
+
+For Kubernetes, a valid kubeconfig file must be provided with cluster
+and auth information. The sync process will look into the default locations
+for both in-cluster and out-of-cluster authentication. If `kubectl` works,
+then the sync program should work.
+
+## Kubernetes to Consul
+
+This sync registers Kubernetes services to the Consul catalog automatically.
+
+This enables discovery and connection to Kubernetes services using native
+Consul service discovery such as DNS or HTTP. This is particularly useful for
+non-Kubernetes nodes. This also causes all discoverable services to be part of
+a central service catalog in Consul for further syncing into alternate
+Kubernetes clusters or other platforms.
+
+### Kubernetes Service Types
+
+Not all Kubernetes services are externally accessible. The sync program by
+default will only sync services of the following types or configurations.
+If a service type is not listed below, then the sync program will ignore that
+service type.
+
+#### NodePort
+
+[NodePort services](https://kubernetes.io/docs/concepts/services-networking/service/#nodeport)
+register a static port that every node in the K8S cluster listens on.
+
+For NodePort services, a Consul service instance will be created for each
+node that has the representative pod running. While Kubernetes configures
+a static port on all nodes in the cluster, this limits the number of service
+instances to be equal to the nodes running the target pods.
+
+The service instances registered will be registered to the K8S node name
+that each instance lives on. This is guaranteed unique by Kubernetes. An
+existing node entry will be used if it is already part of the Consul
+cluster (for example if you're running a client agent on all Kubernetes
+nodes). This allows the normal agent health checks for that node to continue
+working.
+
+#### LoadBalancer
+
+For LoadBalancer services, a single service instance will be registered with
+the external IP of the created load balancer. Because this is already a load
+balancer, only one service instance will be registered with Consul rather
+than registering each individual pod endpoint.
+
+#### External IPs
+
+Any service type may specify an
+"[external IP](https://kubernetes.io/docs/concepts/services-networking/service/#external-ips)"
+configuration. The external IP must be configured by some other system, but
+any service discovery will resolve to this set of IP addresses rather than a
+virtual IP.
+
+If an external IP list is present, a service instance in Consul will be created
+for each external IP. It is assumed that if an external IP is present that it
+is routable and configured by some other system.
+
+### Sync Enable/Disable
+
+By default, all valid services (as explained above) are synced. This default
+can be changed as configuration to the sync process. Syncing can also be
+explicitly enabled or disabled using an annotation:
```

Comment:
Is it possible to explicitly enable sync for an unsupported kubernetes service type, i.e. `ClusterIP`?

Reply:
I think since `ClusterIP` is a virtual IP and choosing it makes the kubernetes service only reachable from within the cluster. This means that nodes/services outside Kubernetes cluster won't be able to reach that service directly anyway, they will have to either use an Ingress controller or other service types like `NodePort` or `LoadBalancer` to access the service. It won't make much sense to sync that service to Consul since it is not discoverable externally unless someone is explicitly **not** trying to use kube-dns internally for communication inside the cluster and wanting to replace kube-dns with Consul DNS, which is unlikely.

Reply:
Not at launch, but its something we can and should consider for future support as we get more use cases, i.e. that'd work totally fine if the kube-proxy is locally available and running even if its external to the K8S cluster, technically. We just have to learn more about the use case.

Reply:
What we'd actually like to see is pulling in the endpoints of the ClusterIP instead of the ClusterIP itself. (Our pod IPs are on a flat network and so routable directly.)

Reply:
That's a totally reasonable ask, and we already have the endpoint syncing functionality in there for `NodePort`. We could add this without huge issue... @roboll I'll try to remember but if you can report this as a feature request on the consul-k8s repo (going OSS in a few hours) that'd be _super_ helpful.

Reply:
@mitchellh Using endpoints under the service (type `ClusterIP`) should work in most clusters with routable pod ips (except clusters running overlay).